File: src/main/java/de/rubixdev/enchanted_shulkers/enchantment/RefillEnchantment.java
package de.rubixdev.enchanted_shulkers.enchantment;

import de.rubixdev.enchanted_shulkers.Mod;
import de.rubixdev.enchanted_shulkers.Utils;
import java.util.List;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Pair;
import net.minecraft.util.collection.DefaultedList;

public class RefillEnchantment extends Enchantment {
    public RefillEnchantment() {
        super(Rarity.RARE, Mod.PORTABLE_CONTAINER_TARGET, new EquipmentSlot[] {
            EquipmentSlot.MAINHAND, EquipmentSlot.OFFHAND
        });
    }

    @Override
    public int getMinPower(int level) {
        return level * 25;
    }

    @Override
    public int getMaxPower(int level) {
        return getMinPower(level) + 50;
    }

    @Override
    public boolean isTreasure() {
        return true;
    }

    private static final Previous previous = new Previous();

    private static class Previous {
        int slot = -1;
        ItemStack mainStack = ItemStack.EMPTY;
        ItemStack offStack = ItemStack.EMPTY;
    }

    public static void onPlayerTick(ServerPlayerEntity player) {
        int currentSlot = player.getInventory().selectedSlot;
        ItemStack currentMainStack = player.getInventory().getMainHandStack();
        ItemStack currentOffStack = player.getInventory().getStack(PlayerInventory.OFF_HAND_SLOT);

        boolean swappedHands = ItemStack.areNbtEqual(previous.mainStack, currentOffStack)
                && ItemStack.areNbtEqual(currentMainStack, previous.offStack);
        boolean wasMainEmptied = previous.mainStack.getCount() > 0 && currentMainStack.isEmpty() && !swappedHands;
        boolean wasOffEmptied = previous.offStack.getCount() > 0 && currentOffStack.isEmpty() && !swappedHands;

        if (currentSlot != previous.slot
                || swappedHands
                || !(wasMainEmptied || ItemStack.canCombine(currentMainStack, previous.mainStack))
                || !(wasOffEmptied || ItemStack.canCombine(currentOffStack, previous.offStack))) {
            previous.slot = currentSlot;
            previous.mainStack = currentMainStack.copy();
            previous.offStack = currentOffStack.copy();
            return;
        }

        if (currentMainStack.getCount() < previous.mainStack.getCount()) {
            refill(
                    player,
                    currentSlot,
                    previous.mainStack,
                    previous.mainStack.getCount() - currentMainStack.getCount());
        } else if (currentOffStack.getCount() < previous.offStack.getCount()) {
            refill(
                    player,
                    PlayerInventory.OFF_HAND_SLOT,
                    previous.offStack,
                    previous.offStack.getCount() - currentOffStack.getCount());
        }
        previous.mainStack = currentMainStack.copy();
        previous.offStack = currentOffStack.copy();
    }

    static void refill(ServerPlayerEntity player, int slot, ItemStack itemType, int amount) {
        List<Pair<Integer, ItemStack>> containerSlots =
                Utils.getContainers(player.getInventory(), Mod.REFILL_ENCHANTMENT);
        for (Pair<Integer, ItemStack> containerSlot : containerSlots) {
            if (amount <= 0) return;
            ItemStack container = containerSlot.getRight();
            DefaultedList<ItemStack> containerInventory = Utils.getContainerInventory(container);

            boolean updateContainer = false;
            for (ItemStack innerStack : containerInventory) {
                if (innerStack.isEmpty()) continue;
                int refilled = tryRefillSlot(player.getInventory(), innerStack, slot, itemType, amount);
                if (refilled > 0) {
                    amount -= refilled;
                    updateContainer = true;
                    if (amount <= 0) break;
                }
            }
            if (updateContainer) Utils.setContainerInventory(container, containerInventory);
        }
    }

    static int tryRefillSlot(PlayerInventory inventory, ItemStack from, int slot, ItemStack itemType, int amount) {
        ItemStack to = inventory.getStack(slot);

        if (!ItemStack.canCombine(itemType, from)) return 0;
        if (!to.isEmpty() && !ItemStack.canCombine(to, from)) return 0;

        int transferCount = Math.min(Math.min(amount, itemType.getMaxCount() - to.getCount()), from.getCount());
        if (transferCount <= 0) return 0;

        if (to.isEmpty()) {
            ItemStack stack = from.copy();
            stack.setCount(transferCount);
            inventory.setStack(slot, stack);
        } else to.increment(transferCount);
        from.decrement(transferCount);
        return transferCount;
    }
}

File: versions/26.2/src/main/kotlin/de/rubixdev/enchantedshulkers/enchantment/RefillEnchantment.kt
package de.rubixdev.enchantedshulkers.enchantment

import de.rubixdev.enchantedshulkers.Mod
import de.rubixdev.enchantedshulkers.Utils
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.item.ItemStack

/**
 * 26.2 port - see SiphonEnchantment.kt's class doc for why this is a plain object now.
 *
 * TODO(26.2): config/ isn't ported yet, so `creativeRefill`/`refillNonStackables`/`refillOffhand`
 * WorldConfig toggles are dropped for now (creative players always allowed to refill; non-
 * stackable refilling off; offhand refilling on, matching the original's defaults).
 */
@Suppress("NAME_SHADOWING")
object RefillEnchantment {
    @JvmStatic
    fun onPlayerTick(
        player: ServerPlayer,
        inventoryOpen: Boolean,
        currentSlot: Int,
        currentMainStack: ItemStack,
        currentOffStack: ItemStack,
        previousSlot: Int,
        previousMainStack: ItemStack,
        previousOffStack: ItemStack,
    ) {
        val allowsRefill = !inventoryOpen
        val swappedHands = ItemStack.isSameItemSameComponents(previousMainStack, currentOffStack) &&
            ItemStack.isSameItemSameComponents(currentMainStack, previousOffStack)
        val wasMainEmptied = previousMainStack.count > 0 && currentMainStack.isEmpty && !swappedHands
        val wasOffEmptied = previousOffStack.count > 0 && currentOffStack.isEmpty && !swappedHands
        val shouldRefillMain = (wasMainEmptied || ItemStack.isSameItemSameComponents(currentMainStack, previousMainStack)) &&
            currentMainStack.count < previousMainStack.count && previousMainStack.isStackable
        val shouldRefillOff = (wasOffEmptied || ItemStack.isSameItemSameComponents(currentOffStack, previousOffStack)) &&
            currentOffStack.count < previousOffStack.count && previousOffStack.isStackable
        val doRefill = allowsRefill && currentSlot == previousSlot && !swappedHands && (shouldRefillMain || shouldRefillOff)

        if (!doRefill) return
        if (shouldRefillMain) {
            refill(player, currentSlot, previousMainStack, previousMainStack.count - currentMainStack.count)
        } else {
            refill(player, Inventory.SLOT_OFFHAND, previousOffStack, previousOffStack.count - currentOffStack.count)
        }
    }

    @JvmStatic
    fun refill(player: ServerPlayer, slot: Int, itemType: ItemStack, amount: Int): Boolean {
        if (amount <= 0) return false
        var didRefill = false
        var remaining = amount
        val refillHolder = Utils.enchantmentHolder(player.registryAccess(), Mod.REFILL_KEY)
        val augmentHolder = Utils.enchantmentHolder(player.registryAccess(), Mod.AUGMENT_KEY)
        val containerSlots = Utils.getContainers(player, refillHolder)
        for (container in containerSlots) {
            val containerInventory = Utils.getContainerInventory(container, augmentHolder)
            var updateContainer = false
            for (innerStack in containerInventory) {
                if (innerStack.isEmpty) continue
                val refilled = tryRefillSlot(player.inventory, innerStack, slot, itemType, remaining)
                if (refilled > 0) {
                    remaining -= refilled
                    updateContainer = true
                    if (remaining <= 0) break
                }
            }
            if (updateContainer) {
                didRefill = true
                Utils.setContainerInventory(container, containerInventory)
            }
        }
        return didRefill
    }

    private fun tryRefillSlot(inventory: Inventory, from: ItemStack, slot: Int, itemType: ItemStack, amount: Int): Int {
        val to = inventory.getItem(slot)

        if (!ItemStack.isSameItemSameComponents(itemType, from)) return 0
        if (!to.isEmpty && !ItemStack.isSameItemSameComponents(to, from)) return 0

        val transferCount = minOf(amount, itemType.maxStackSize - to.count, from.count)
        if (transferCount <= 0) return 0

        if (to.isEmpty) {
            val stack = from.copy()
            stack.count = transferCount
            inventory.setItem(slot, stack)
        } else {
            to.grow(transferCount)
        }
        from.shrink(transferCount)
        return transferCount
    }
}
